File: src/pypaginate/filtering/operators.py
"""Filter operators for in-memory predicate evaluation."""

from __future__ import annotations

from collections.abc import Iterable
from typing import Any, Protocol, runtime_checkable

from pypaginate.domain.exceptions import FilterError, FilterValidationError
from pypaginate.filtering.like import match_ilike, match_like
from pypaginate.filtering.regex import compile_pattern


@runtime_checkable
class Operator(Protocol):
    """Protocol for filter operators."""

    @staticmethod
    def evaluate(field_value: object, spec_value: object) -> bool: ...


class Eq:
    __slots__ = ()

    @staticmethod
    def evaluate(field_value: object, spec_value: object) -> bool:
        return field_value == spec_value


class Ne:
    __slots__ = ()

    @staticmethod
    def evaluate(field_value: object, spec_value: object) -> bool:
        return field_value != spec_value


class Gt:
    __slots__ = ()

    @staticmethod
    def evaluate(field_value: object, spec_value: object) -> bool:
        return bool(field_value > spec_value)  # type: ignore[operator]


class Gte:
    __slots__ = ()

    @staticmethod
    def evaluate(field_value: object, spec_value: object) -> bool:
        return bool(field_value >= spec_value)  # type: ignore[operator]


class Lt:
    __slots__ = ()

    @staticmethod
    def evaluate(field_value: object, spec_value: object) -> bool:
        return bool(field_value < spec_value)  # type: ignore[operator]


class Lte:
    __slots__ = ()

    @staticmethod
    def evaluate(field_value: object, spec_value: object) -> bool:
        return bool(field_value <= spec_value)  # type: ignore[operator]


class In:
    __slots__ = ()

    @staticmethod
    def evaluate(field_value: object, spec_value: object) -> bool:
        return field_value in spec_value  # type: ignore[operator]


class NotIn:
    __slots__ = ()

    @staticmethod
    def evaluate(field_value: object, spec_value: object) -> bool:
        return field_value not in spec_value  # type: ignore[operator]


class Contains:
    __slots__ = ()

    @staticmethod
    def evaluate(field_value: object, spec_value: object) -> bool:
        return str(spec_value) in str(field_value)


class StartsWith:
    __slots__ = ()

    @staticmethod
    def evaluate(field_value: object, spec_value: object) -> bool:
        return str(field_value).startswith(str(spec_value))


class EndsWith:
    __slots__ = ()

    @staticmethod
    def evaluate(field_value: object, spec_value: object) -> bool:
        return str(field_value).endswith(str(spec_value))


class Like:
    __slots__ = ()

    @staticmethod
    def evaluate(field_value: object, spec_value: object) -> bool:
        return match_like(str(field_value), str(spec_value))


class ILike:
    __slots__ = ()

    @staticmethod
    def evaluate(field_value: object, spec_value: object) -> bool:
        return match_ilike(str(field_value), str(spec_value))


class Between:
    __slots__ = ()

    @staticmethod
    def evaluate(field_value: object, spec_value: object) -> bool:
        low, high = _unpack_pair(spec_value)
        return bool(low <= field_value <= high)


class IsNull:
    __slots__ = ()

    @staticmethod
    def evaluate(field_value: object, _spec_value: object) -> bool:
        return field_value is None


class IsNotNull:
    __slots__ = ()

    @staticmethod
    def evaluate(field_value: object, _spec_value: object) -> bool:
        return field_value is not None


class Regex:
    __slots__ = ()

    @staticmethod
    def evaluate(field_value: object, spec_value: object) -> bool:
        pattern = str(spec_value)
        try:
            compiled = compile_pattern(pattern)
            return bool(compiled.search(str(field_value)))
        except Exception as exc:
            raise FilterError(
                f"Invalid regex pattern: '{pattern}'",
                details={"pattern": pattern, "error": str(exc)},
            ) from exc


class Empty:
    __slots__ = ()

    @staticmethod
    def evaluate(field_value: object, _spec_value: object) -> bool:
        return field_value is None or field_value in ("", [])


class NotEmpty:
    __slots__ = ()

    @staticmethod
    def evaluate(field_value: object, _spec_value: object) -> bool:
        return field_value is not None and field_value not in ("", [])


class Exists:
    __slots__ = ()

    @staticmethod
    def evaluate(_field_value: object, _spec_value: object) -> bool:
        return True


def _unpack_pair(value: object) -> tuple[Any, Any]:
    """Unpack a two-element sequence for Between."""
    if not isinstance(value, Iterable):
        raise FilterValidationError(
            "Between requires a two-element sequence",
            details={"value": repr(value), "type": type(value).__name__},
        )
    seq = list(value)
    if len(seq) != 2:
        raise FilterValidationError(
            "Between requires exactly 2 elements",
            details={"value": repr(value), "length": len(seq)},
        )
    return seq[0], seq[1]

__all__ = [
    "Between",
    "Contains",
    "Empty",
    "EndsWith",
    "Eq",
    "Exists",
    "Gt",
    "Gte",
    "ILike",
    "In",
    "IsNotNull",
    "IsNull",
    "Like",
    "Lt",
    "Lte",
    "Ne",
    "NotEmpty",
    "NotIn",
    "Operator",
    "Regex",
    "StartsWith",
]

File: crates/core/src/filter/operators.rs
//! Evaluation of the comparison, membership, range, and null/empty operators.
//!
//! The string operators (`contains`/`starts_with`/`ends_with`/`like`/`ilike`)
//! and `regex` are compiled with precomputed state in [`super::compile_spec`]
//! and never reach this function.

use std::cmp::Ordering;

use super::FilterOp;
use crate::coerce;
use crate::error::{CoreError, Result};
use crate::value::Value;

pub(super) fn eval_op(op: FilterOp, field: &Value, spec: &Value) -> Result<bool> {
    match op {
        FilterOp::Eq => Ok(coerce::eq(field, spec)),
        FilterOp::Ne => Ok(!coerce::eq(field, spec)),
        FilterOp::Gt => cmp(field, spec, |o| o == Ordering::Greater),
        FilterOp::Gte => cmp(field, spec, |o| o != Ordering::Less),
        FilterOp::Lt => cmp(field, spec, |o| o == Ordering::Less),
        FilterOp::Lte => cmp(field, spec, |o| o != Ordering::Greater),
        FilterOp::In => member(field, spec),
        FilterOp::NotIn => member(field, spec).map(|m| !m),
        FilterOp::Between => between(field, spec),
        FilterOp::IsNull => Ok(field.is_null()),
        FilterOp::IsNotNull => Ok(!field.is_null()),
        FilterOp::Empty => Ok(is_empty(field)),
        FilterOp::NotEmpty => Ok(!is_empty(field)),
        FilterOp::Exists => Ok(true),
        FilterOp::Contains
        | FilterOp::StartsWith
        | FilterOp::EndsWith
        | FilterOp::Like
        | FilterOp::ILike
        | FilterOp::Regex => {
            unreachable!("string/regex operators are compiled separately")
        }
    }
}

fn cmp(field: &Value, spec: &Value, pred: impl Fn(Ordering) -> bool) -> Result<bool> {
    match coerce::compare(field, spec) {
        Some(ordering) => Ok(pred(ordering)),
        None => Err(CoreError::Filter {
            message: format!("{field:?} and {spec:?} are not order-comparable"),
        }),
    }
}

fn member(field: &Value, spec: &Value) -> Result<bool> {
    match spec {
        Value::List(items) => Ok(items.iter().any(|e| coerce::eq(field, e))),
        // `x in "string"` is a substring test in Python.
        Value::Str(haystack) => Ok(haystack.contains(coerce::to_py_str(field).as_str())),
        // `key in dict` is key membership.
        Value::Map(map) => Ok(map.contains_key(coerce::to_py_str(field).as_str())),
        other => Err(CoreError::Filter {
            message: format!("argument of type {other:?} is not iterable"),
        }),
    }
}

fn between(field: &Value, spec: &Value) -> Result<bool> {
    let pair = match spec {
        Value::List(v) if v.len() == 2 => v,
        Value::List(_) => {
            return Err(CoreError::Filter {
                message: "Between requires exactly 2 elements".to_owned(),
            })
        }
        _ => {
            return Err(CoreError::Filter {
                message: "Between requires a two-element sequence".to_owned(),
            })
        }
    };
    let (low, high) = (&pair[0], &pair[1]);
    // `low <= field <= high`, short-circuiting like Python's chained comparison.
    match coerce::compare(low, field) {
        Some(Ordering::Greater) => return Ok(false),
        Some(_) => {}
        None => return Err(not_comparable()),
    }
    match coerce::compare(field, high) {
        Some(ordering) => Ok(ordering != Ordering::Greater),
        None => Err(not_comparable()),
    }
}

fn not_comparable() -> CoreError {
    CoreError::Filter {
        message: "Between bounds are not order-comparable with the field".to_owned(),
    }
}

/// `field is None or field in ("", [])`.
fn is_empty(field: &Value) -> bool {
    match field {
        Value::Null => true,
        Value::Str(s) => s.is_empty(),
        Value::List(l) => l.is_empty(),
        _ => false,
    }
}
